### PR TITLE
Domain Adversarial Classifier, Loss, Trainer and example 

### PR DESCRIPTION
**Pull request by ckarouzos**
It's just a first version.
Some Issues to be discussed (@georgepar ): 

- Use of both Loss and Accuracy as metrics.

- Keeping 2 separate losses for heads, and merge them only for the encoder.

### CODE REVIEW

File: slp/modules/daclassifier.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+class DALoss(nn.Module):
+    def __init__(self, loss_fn_cl, loss_fn_d):
+        super(DALoss, self).__init__()
+        self.loss_fn_cl = loss_fn_cl
+        self.loss_fn_d = loss_fn_d
+
+    def forward(self, pred, targets, domain_pred, domain_targets):
+        loss_cl = self.loss_fn_cl(pred, targets)
+        loss_d = self.loss_fn_d(domain_pred, domain_targets)
+        return loss_cl-loss_d #NOTSURE
```
Comment (georgepar) [inline slp/modules/daclassifier.py]: add the losses and use grad inversion. this is no good

File: examples/imdb_smt.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+class DatasetWrapper(Dataset):
+    def __init__(self, dataset, name):
+        self.dataset = dataset
+        self.name = 0 if name == 'imdb' else 1
+        self.label = 'sentiment' if name == 'imdb' else 'label'
+        self.transforms = []
+        self.label_encoder = (LabelEncoder()
+                              .fit([d[self.label] for d in dataset]))
+
+    def map(self, t):
```
Comment (georgepar) [inline examples/imdb_smt.py:39]: there is also LambdaDataset under slp/data/datasets.py so you don't need to implement map all the time

File: slp/trainer/trainer.py [hunk 3]
```diff
@@ -290,3 +290,56 @@ def get_predictions_and_targets(
...
+    def eval_step(
+            self: TrainerType,
+            engine: Engine,
+            batch: List[torch.Tensor]) -> Tuple[torch.Tensor, ...]:
+        self.model.eval()
+        with torch.no_grad():
+            y_pred, targets, d_pred, domains = self.get_predictions_and_targets(batch)
+            d = {'domain_pred'  : d_pred,
+		 'domain_targets'  : domains}
+            return y_pred, targets, d
```
Comment (georgepar) [inline slp/trainer/trainer.py:344]: This will break metrics calculation I think

Comment (ckarouzos) [reply slp/trainer/trainer.py:344]: It's the way to use IGNITE.METRICS.LOSS for a generic loss with more arguments. 
https://pytorch.org/ignite/_modules/ignite/metrics/loss.html

But it is not working for other metrics (e.g. Accuracy).

**Comment by ckarouzos** [comment]
Actually something goes completely wrong. Need to discuss.

**Comment by ckarouzos** [comment]
@georgepar Let's check again.
